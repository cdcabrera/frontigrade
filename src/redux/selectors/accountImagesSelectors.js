import { createSelector } from 'reselect';
import _cloneDeep from 'lodash/cloneDeep';

/**
 * Account Image detail selector
 */
const imageDetail = (state, props) => state.accountImageEdit[props.id];

const imageDetailSelector = createSelector(
  [imageDetail],
  detail => detail
);

const makeImageDetailSelector = () => imageDetailSelector;

/**
 * Account Images view selector
 */
const imagesView = state => state.accountImages.view;

const imagesImageDetail = state => state.accountImageEdit;

const imagesImageAccount = state => state.accountImages.account;

const imagesViewSelector = createSelector(
  [imagesView, imagesImageDetail, imagesImageAccount],
  (imgsView, imgsImageDetail, imgsImageAccount) => {
    const detailKeys = Object.keys(imgsImageDetail);

    if (!detailKeys.length) {
      return {
        ...imgsView,
        account: imgsImageAccount
      };
    }

    const updatedImgs = _cloneDeep(imgsView.images);

    detailKeys.forEach(value => {
      const index = updatedImgs.findIndex(image => Number.parseInt(value, 10) === image.id);
      updatedImgs[index] = {
        ...updatedImgs[index],
        ...imgsImageDetail[value].image
      };
    });

    return {
      ...imgsView,
      ...{
        images: updatedImgs
      },
      account: imgsImageAccount
    };
  }
);

const makeImagesViewSelector = () => imagesViewSelector;

const accountImagesSelectors = {
  imagesView: imagesViewSelector,
  makeImagesView: makeImagesViewSelector,
  imageDetail: imageDetailSelector,
  makeImageDetail: makeImageDetailSelector
};

export {
  accountImagesSelectors as default,
  accountImagesSelectors,
  imagesViewSelector,
  makeImagesViewSelector,
  imageDetailSelector,
  makeImageDetailSelector
};
