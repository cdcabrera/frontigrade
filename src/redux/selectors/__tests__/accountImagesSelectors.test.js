import accountImagesSelectors from '../accountImagesSelectors';

describe('AccountImagesSelectors', () => {
  it('Should return specific selectors', () => {
    expect(accountImagesSelectors).toMatchSnapshot('selectors');
  });

  it('Should filter the state with imageDetail', () => {
    const state = {
      accountImageEdit: {
        2: {
          test: 'success'
        }
      }
    };
    const props = {
      id: 2
    };

    expect(accountImagesSelectors.imageDetail(state, props)).toMatchSnapshot('imageDetail');
  });

  it('Should filter the state with imagesView', () => {
    const state = {
      accountImages: {
        view: {
          images: [{ id: 2 }]
        }
      },
      accountImageEdit: {}
    };

    expect(accountImagesSelectors.imagesView(state)).toMatchSnapshot('imagesView');

    const updatedState = {
      accountImages: {
        view: {
          images: [{ id: 2 }]
        }
      },
      accountImageEdit: {
        2: {
          image: {
            test: 'success'
          }
        }
      }
    };

    expect(accountImagesSelectors.imagesView(updatedState)).toMatchSnapshot('imagesViewUpdated');
  });
});
