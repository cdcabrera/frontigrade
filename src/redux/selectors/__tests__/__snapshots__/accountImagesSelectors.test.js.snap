// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountImagesSelectors Should filter the state with imageDetail: imageDetail 1`] = `
Object {
  "test": "success",
}
`;

exports[`AccountImagesSelectors Should filter the state with imagesView: imagesView 1`] = `
Object {
  "account": undefined,
  "images": Array [
    Object {
      "id": 2,
    },
  ],
}
`;

exports[`AccountImagesSelectors Should filter the state with imagesView: imagesViewUpdated 1`] = `
Object {
  "account": undefined,
  "images": Array [
    Object {
      "id": 2,
      "test": "success",
    },
  ],
}
`;

exports[`AccountImagesSelectors Should return specific selectors: selectors 1`] = `
Object {
  "imageDetail": [Function],
  "imagesView": [Function],
  "makeImageDetail": [Function],
  "makeImagesView": [Function],
}
`;
